### DISCUSSION TITLE
Is excess time spent in preempt_count_sub() a profiling artifact?

### DISCUSSION THREAD

**Discussion by jfrantzius** [async-profiler/async-profiler#707] [Q&A] [answered]
Hi,
in a [Quarkus discussion on profiling a benchmark](https://github.com/quarkusio/quarkus/discussions/30231#discussioncomment-4648573), I observed some excess time being spent in the `preempt_count_sub()` Kernel function during `Unsafe.unpark()` (in a Linux v5.10.124 aarch64 image running on Mac M1, with profiling running for 8 minutes):
![image](https://user-images.githubusercontent.com/1823661/211647347-ad2fd72a-4ba8-4cbb-a4eb-52046e336c03.png)
Given that `preempt_count_sub()` is rather short (only inlined code and a macro), and probably supposed to execute very quickly, this seems odd to me.

Now I wonder whether this could somehow be an artifact of the profiliing itself?

Also I wondered what the numbers 5,967 and 5,951 next to `preempt_count_sub()` mean exactly? I guess one of them is the number of invocations, but unfortunately I wasn't able to find documentation on the tree output format.

Thanks in advance for any pointers!

**Comment by apangin** [comment]
I suppose you run profiler in `-e cpu` mode, not `-e cycles`?
If so, kernel stacktraces, especially related to context switching, may appear imprecise. A kind of "safepoint bias" (actually, ["IRQ bias"](https://www.spinics.net/lists/linux-perf-users/msg08711.html)) in the kernel space :)

> Also I wondered what the numbers 5,967 and 5,951 next to preempt_count_sub() mean exactly?

The number of samples: total and self, i.e. how many times this stack trace was seen during profiling.

**Comment by jfrantzius** [comment]
Thanks for your answer @apangin ! Yes indeed I used `-e cpu`. 

I now tried `-e cycles` with both `sysctl kernel.perf_event_paranoid=1` and `sysctl kernel.perf_event_paranoid=-1` (which according to [kernel docs](https://stackoverflow.com/questions/51911368/what-restriction-is-perf-event-paranoid-1-actually-putting-on-x86-perf) seemed the most permissive) , but unfortunately I always get this when running in a terminal opened manually in the Docker container:

```
# ./async-profiler/profiler.sh -l -d 480 -e cycles -o tree -f dump4.html --fdtransfer jps
[WARN] perf_event_open for TID 11 failed: No such file or directory
[WARN] perf_event_open for TID 12 failed: No such file or directory
[WARN] perf_event_open for TID 13 failed: No such file or directory
[WARN] perf_event_open for TID 14 failed: No such file or directory
[..]
[WARN] perf_event_open for TID 56 failed: No such file or directory
[WARN] perf_event_open for TID 76 failed: No such file or directory
[ERROR] Perf events unavailable
```

Do you per chance know what else I should be doing to enable cycles profiling?

**Reply by apangin** [reply]
Probably, you run on a virtual machine that does not expose HW performance counters.
EDIT: Oh, I see you mentioned Mac M1. I'm not sure if it provides userspace access to hardware performance counters at all.

**Comment by jfrantzius** [comment, marked as answer]
Alright, I re-ran the test with Docker on a Linux x86 host, and the phenomenon simply doesn't occur there.

So the conclusion for me is: running and profiling benchmarks via Docker on Mac can be very misleading...
